python/ray/data/datasource/bigquery_datasource.py

```diff
@@ -0,0 +1,185 @@
+import logging
+import os
+import tempfile
+import time
+import uuid
+from typing import Any, Dict, List, Optional
+
+import pyarrow.parquet as pq
+from google.api_core import exceptions
+from google.cloud import bigquery, bigquery_storage
+from google.cloud.bigquery_storage import types
+
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+from ray.types import ObjectRef
+
+logger = logging.getLogger(__name__)
+
+
+class _BigQueryDatasourceReader(Reader):
+    def __init__(
+        self,
+        project_id: str,
+        dataset: Optional[str] = None,
+        query: Optional[str] = None,
+        parallelism: Optional[int] = -1,
+        **kwargs: Optional[Dict[str, Any]],
+    ):
+        self._project_id = project_id
+        self._dataset = dataset
+        self._query = query
+        self._kwargs = kwargs
+
+        if query is not None and dataset is not None:
+            raise ValueError(
+                "Query and dataset kwargs cannot both be provided"
+                + " (must be mutually exclusive)."
```

Comment:
I don't think you need the `+ ` here?

```suggestion
                " (must be mutually exclusive)."
```